vllm/config.py

```diff
@@ -669,26 +674,51 @@ def get_num_attention_heads(self,
         num_heads = getattr(self.hf_text_config, "num_attention_heads", 0)
         return num_heads // parallel_config.tensor_parallel_size
 
-    def get_num_layers(self, parallel_config: "ParallelConfig") -> int:
+    def get_layers_start_end_indices(
+            self, parallel_config: "ParallelConfig") -> Tuple[int, int]:
         from vllm.distributed.utils import get_pp_indices
         total_num_hidden_layers = getattr(self.hf_text_config,
                                           "num_hidden_layers", 0)
         pp_rank = parallel_config.rank // parallel_config.tensor_parallel_size
         pp_size = parallel_config.pipeline_parallel_size
         start, end = get_pp_indices(total_num_hidden_layers, pp_rank, pp_size)
+        return start, end
+
+    def get_num_layers(self, parallel_config: "ParallelConfig") -> int:
+        start, end = self.get_layers_start_end_indices(parallel_config)
         return end - start
 
-    def get_num_attention_layers(self,
-                                 parallel_config: "ParallelConfig") -> int:
-        if self.is_attention_free:
+    def get_num_layers_by_block_type(
+        self,
+        parallel_config: "ParallelConfig",
+        block_type: LayerBlockType = LayerBlockType.attention,
+    ) -> int:
+        # This function relies on 'layers_block_type' in hf_config,
+        # for hybrid/attention-free models w/o this attribute,
+        # we will need to have workarounds like so
+        attn_block_type = block_type == LayerBlockType.attention
+        is_full_attn_model = not self.is_hybrid and not self.is_attention_free
+
+        if self.is_attention_free and attn_block_type:
+            return 0
+        if is_full_attn_model and not attn_block_type:
             return 0
 
-        num_layers = self.get_num_layers(parallel_config)
+        start, end = self.get_layers_start_end_indices(parallel_config)
 
         # Transformers supports layers_block_type @property
-        layers = getattr(self.hf_config, "layers_block_type",
-                         ["attention"] * num_layers)
-        return len([t for t in layers if t == "attention"])
+        layers_block_type_value = getattr(self.hf_config, "layers_block_type",
+                                          None)
+        if layers_block_type_value is None:
+            if not self.is_hybrid:
+                return end - start
+            raise ValueError("The model is an hybrid without a"
+                             "layers_block_type in the hf_config,"
+                             "cannot determine the num of "
+                             f"{block_type.value} layers")
+
+        return sum(t == block_type.value
+                   for t in layers_block_type_value[start:end])
```

Comment:
My last review comment was a little off because I thought this was already handling full-attention models before getting to the code that handles the hybrid case.

I think it would be simpler if we explicitly handle the three cases we'll run into: attention-free, transformers, and hybrid models:

```
attn_block_type = block_type == LayerBlockType.attention
is_transformer = not self.is_hybrid and not self.is_attention_free

if is_transformer:
    # Handle the basic case first
    return end - start if attn_block_type else 0
elif is_attention_free:
    # Attention free
    # Note that this code assumes there is only one type of attention-free block type.
    return 0 if attn_block_type else end - start
else:
    # Hybrid model
    layers_block_type_value = getattr(self.hf_config, "layers_block_type",
                                          None)
        if layers_block_type_value is None:
            raise ValueError("The model is an hybrid without a"
                             "layers_block_type in the hf_config,"
                             "cannot determine the num of "
                             f"{block_type.value} layers")

        return sum(t == block_type.value
                   for t in layers_block_type_value[start:end])
```

Reply:
Yeah, looks better this way, Fixed it accordingly